### PR TITLE
Gacha card creation [collab branch]

### PR DESCRIPTION
**Pull request by austinpray**
[💹 PR Code Coverage](https://github.com/austinpray/kaori/pull/66#issuecomment-643504824)

A user should be able to

- `kaori create card` to create a card in the database
- `kaori show card {name}` to search for a specific owned card
- `kaori show cards` to show owned cards
- utilities like help and price list

### blockers

- [x] upload user images to google cloud

### important but not blocking 

- [x] ~~DRY difference between engine cards and card models~~ moved to #101 
- [x] behavior unit tests for cursed  next_card_creation_step function

* * * 

## Features split from this PR

- [make DI aware of ABCs](https://github.com/austinpray/kaori/pull/69)
- [proper integration testing](https://github.com/austinpray/kaori/pull/75)
- [image upload skill](https://github.com/austinpray/kaori/pull/91)

## Misc PRs split from this one

- https://github.com/austinpray/kaori/pull/68
- https://github.com/austinpray/kaori/pull/77
- https://github.com/austinpray/kaori/pull/78
- https://github.com/austinpray/kaori/pull/79
- https://github.com/austinpray/kaori/pull/84
- https://github.com/austinpray/kaori/pull/86
- https://github.com/austinpray/kaori/pull/89

### CODE REVIEW

**Comment by austinpray** [comment]
@ridhoq here's current demo https://drive.google.com/file/d/1vrL9VmNorPzUq64UJ4sphqyWaQU1zoIg/view

todos blocking launch are updated in the OP

**Comment by ridhoq** [comment]
Let's gooooooo!!!

In your demo, what's happening for the image upload? (since you've listed the GCP image upload as a blocker)

**Comment by austinpray** [comment]
> In your demo, what's happening for the image upload? (since you've listed the GCP image upload as a blocker)

@ridhoq See 

https://github.com/austinpray/kaori/blob/ab876732e6effa2555088ab0c4ecc21550403ef3/kaori/skills/file_uploads.py#L9-L13

and 

https://github.com/austinpray/kaori/blob/ab876732e6effa2555088ab0c4ecc21550403ef3/kaori/worker/__init__.py#L53-L58

which is injected 

https://github.com/austinpray/kaori/blob/ab876732e6effa2555088ab0c4ecc21550403ef3/kaori/plugins/gacha/commands/creation.py#L86

If you provide google creds it will use a (not yet implemented) google storage FileUploader. Otherwise: if you are on dev just fall back to downloading the images to your harddrive and exposing them via docker-compose and ngrok.

If you wanna take a stab at implemented that it should be easy :eyes:. There's a manual e2e test here https://github.com/austinpray/kaori/blob/gacha-card-create/test-e2e/slack_file_download.py. You could copy paste this e2e test to like slack_file_download_gcp.py and hack on it without worrying about all the card creation stuff

File: kaori/plugins/gacha/engine/core/card.py [hunk 1]
```diff
@@ -137,44 +141,47 @@ def detect_standoff(a: Card, b: Card, debug: bool = False) -> bool:
...
-
-            nature_table.add_row(style_row(k, k), style_row(k, v))
-            stat_table.add_row(style_row(k, self.combat_strat.natures[k].boosts), style_row(k, stat))
-
-        table.add_row(nature_table, stat_table)
-        return table
+    def to_rich_table(self, title_style="bold underline yellow") -> str:
+        return self.to_markdown()
+
+    # def to_rich_table(self, title_style="bold underline yellow") -> Table:
```
Comment (ridhoq) [inline kaori/plugins/gacha/engine/core/card.py:147]:
```suggestion
    # Move this to a separate function as per https://github.com/austinpray/kaori/issues/67
    # def to_rich_table(self, title_style="bold underline yellow") -> Table:
```

File: kaori/plugins/gacha/commands/creation.py [hunk 2]
```diff
@@ -0,0 +1,315 @@
...
+                primary_nature, secondary_nature = matches.group(1, 2)
+                if primary_nature == secondary_nature:
+                    replies.append('You have to choose two different natures.')
+                else:
+                    card.primary_nature = primary_nature.lower()
+                    card.secondary_nature = secondary_nature.lower()
+                    replies.append(':+1:')
+                    cursor = 'query_rarity'
+            else:
+                replies.append('You need to specify some natures.')
```
Comment (ridhoq) [inline kaori/plugins/gacha/commands/creation.py:241]: Perhaps there should be a check here to if only one nature was specified (atleast a diff message if user only provided one)

File: kaori/plugins/gacha/commands/creation.py [hunk 3]
```diff
@@ -0,0 +1,315 @@
...
+
+    if cursor.startswith('set_'):
+        if len(extract_mentions(user_input)):
+            replies.append('Mentions are not allowed in gacha card text')
+            return card, replies
+
+        if cursor == 'set_name':
+            card.set_name(user_input)
+
+            num_dupes = session \
```
Comment (ridhoq) [inline kaori/plugins/gacha/commands/creation.py:211]: Would like to see this in its own fn for testing purposes

File: kaori/plugins/gacha/commands/creation.py [hunk 4]
```diff
@@ -0,0 +1,315 @@
...
+
+    return card
+
+
+# This is essentially a state machine.
+# Right now this is spaghetti code for prototyping purposes.
+# TODO: Once the functionality is totally built out we can break this into multiple functions etc.
+# TODO: oh god break this into some functions please, this should be a dispatcher or sthn
+# TODO: this function is now officially out of control
+def next_card_creation_step(card: Card,
```
Comment (ridhoq) [inline kaori/plugins/gacha/commands/creation.py:188]: Yeah not sure how you can write tests for this as it is

File: kaori/plugins/gacha/models/Card.py [hunk 5]
```diff
@@ -79,19 +88,32 @@ def validate_name(name: str) -> Tuple[bool, Optional[str]]:
...
         }
 
+    def price(self) -> int:
+        return self.rarity_prices()[self.rarity_mapping[self.rarity]]
+
+    def roll_stats(self):
+        card = EngineCard.generate(self.name,
+                                   self.rarity_mapping[self.rarity],
+                                   [self.primary_nature, self.secondary_nature])
+        # todo: smelly
```
Comment (ridhoq) [inline kaori/plugins/gacha/models/Card.py:109]: yuge

**Comment by ridhoq** [comment]
First pass

File: kaori/plugins/gacha/commands/creation.py [hunk 6]
```diff
@@ -0,0 +1,315 @@
...
+            if re.search(r'yes+|yep+|ye+|yeah+', user_input, re.IGNORECASE):
+                card.published = True
+                cursor = 'done'
+            # todo: make utility for capturing english negatives
+            elif re.search(r'no+|nope+|', user_input, re.IGNORECASE):
+                replies.append("Okay. If you change your mind:\n"
+                               "• Reply with `@kaori yes` to accept the card\n"
+                               "• Reply with `@kaori start over` to trash your current card")
+            else:
+                replies.append("I don't understand: yes or no?")
```
[the remark this answers is not in the document]
Comment (austinpray) [reply kaori/plugins/gacha/commands/creation.py:265]: @ridhoq 

```suggestion
                replies.append("I don't understand. Would you like to create this card? yes or no?")
```

?

[the remark this answers is not in the document]
Comment (ridhoq) [reply kaori/plugins/gacha/commands/creation.py:265]: "I didn't understand - did you say yes or no?"

[the remark this answers is not in the document]
Comment (ridhoq) [reply kaori/plugins/gacha/commands/creation.py:265]: Total nit tbh - take it or leave it

File: kaori/plugins/gacha/models/Card.py (hunk 5 above)
Comment (austinpray) [reply kaori/plugins/gacha/models/Card.py:109]:
```suggestion
        # todo: smelly: we should not have to repeat ourselves here
        # body: This seems to indicate the card model and the engine card are not cohesive
```

Comment (ridhoq) [reply kaori/plugins/gacha/models/Card.py]: dammit thought it was a new attribute lol

File: kaori/plugins/gacha/commands/creation.py (hunk 4 above)
Comment (austinpray) [reply kaori/plugins/gacha/commands/creation.py:188]: Seems super easy to integration test

File: .github/workflows/ci.yml [hunk 7]
```diff
@@ -32,7 +34,7 @@ jobs:
       env:
         CACHE_FROM_IMAGE: us.gcr.io/${{secrets.GCLOUD_PROJECT_ID}}/kaori/kaori:master-latest
     - name: Run the tests
-      run: ./tests/run.sh
+      run: ./scripts/run.sh
```
Comment (austinpray) [inline .github/workflows/ci.yml:37]:
```suggestion
      run: ./tests/run.sh
```
oops this was a bad merge

**Comment by austinpray** [comment]
Merging blocked due to not enough test coverage: blessed

**Comment by austinpray** [comment]
Note that I uninstalled that TODO bot due to its handling of this thread lmao

File: kaori/plugins/gacha/models/Image.py [hunk 8]
```diff
@@ -22,17 +26,21 @@ class Image(Base):
...
+
+        if not file:
+            raise RuntimeError('no file in message')
 
         user = User.get_by_slack_id(session, message.user)
         if not user:
             raise UserNotFound('no user, try refreshing users')
 
-        url = 'https://storage.googleapis.com/img.kaori.io/static/miku.jpg'
+        name, downloaded_file = download_slack_file(file['id'], slack_client=slack_adapter.client)
```
Comment (austinpray) [inline kaori/plugins/gacha/models/Image.py:42]: as per @ridhoq we should be validating a maximum file size here before proceeding

File: kaori/support/slack_files.py [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+from mimetypes import guess_type
+from pprint import pprint
+from tempfile import SpooledTemporaryFile
+from typing import Tuple
+from urllib.parse import urlparse
+
+import requests
+
+
+def download_slack_file(file_id: str, slack_client) -> Tuple[str, SpooledTemporaryFile]:
```
Comment (austinpray) [inline kaori/support/slack_files.py:12]: as per @ridhoq we should provide a `max_size` param here that raises an exception if the size is exceeded to guard against 1GB files being dumped maliciously

File: kaori/plugins/gacha/models/Image.py (hunk 8 above)
Comment (austinpray) [reply kaori/plugins/gacha/models/Image.py:42]: moved to #99

File: kaori/support/slack_files.py (hunk 9 above)
Comment (austinpray) [reply kaori/support/slack_files.py:12]: moved to #99

**Pull request merged**